state-sync/storage-service/server/src/lib.rs

```diff
@@ -65,6 +72,134 @@ impl Error {
     }
 }
 
+/// A subscription for data received by a client
+pub struct DataSubscriptionRequest {
```

Comment:
On a high level, can you clarify how the subscription model will co-exists with existing pull-based model? i.e., will client subscribe to the server as well as actively pull from the same?

Reply:
My mental model is that a client will pull data while they know they're behind (e.g., the Aptos Data Poller tells them the highest version know by the peers). As soon as they catch-up (i.e., they're now at the highest known version), they'll make one of these requests (so that if new data comes in, they'll receive it immediately). The poller will likely see that new data is there too (depending on the poll frequency), but at this point they'll have it. This is essentially what SSv1 also does today. See my response above about why this isn't a "pure" continuous subscription model.

Reply:
Can you clarify if the clients subscribe to one server on multiple? If one sever, how would they deal is a slow/crashing/malicious server?

Reply:
One server. If the server is slow/crashing, they'll get a timeout and the client will decide what to do (e.g., subscribe or pull directly). If it's malicious, the response will fail verification (where it fails verification depends on what is wrong with the response). If a response fails verification, the client will decide what to do (as above) 😄

Reply:
Thanks my question was around how do clients select a server to subscribe (or resubscribe) to - if they keep re-selecting the same server for subscription, then if the server is slow, it will increase overall state-sync latency. There can also be servers, which falsely advertise latest data and client realizes this only after pulling the data - thus these mailicious servers can create a point of attack slowing down the overall state-sync. May be some sort of reputation based server selection is what we are looking for but I am not sure how complex that will be. In any case, this is slightly orthogonal to this PR as we don't have the client implementation yet, so happy to table this discussion for client change :)

Reply:
> my question was around how do clients select a server to subscribe (or resubscribe) to - if they keep re-selecting the same server for subscription, then if the server is slow, it will increase overall state-sync latency. 

Clients will only subscribe to prioritized peers, e.g., those that are upstream. Remember, we're optimizing for the happy path here. There's little we can actually do if malicious peers wish to increase latency. There's a million different attacks for that. An obvious one being to "behave well" until the exact moment they wish to mount the attack.

> There can also be servers, which falsely advertise latest data and client realizes this only after pulling the data - thus these mailicious servers can create a point of attack slowing down the overall state-sync. May be some sort of reputation based server selection is what we are looking for

Yep. We already have a very primitive port of the scoring system from state sync v1. When a client sends enough malicious/bad/incorrect responses, it is ignored. But, the scoring system needs a serious overhaul. It has a load of deficiencies. An obvious one being that peers are simply ignored and not replaced.

My plan going forward is to add topological awareness and more complex insights into the state of our peers. One of the results of that is we'll have more in-depth information to decide which peers to prioritize and also how to better score peers. But, one step at a time 😄